[TP] Introduce Sequence Parallel Style for Laynorm/RMSNorm/Dropout

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121295
* #121294

As titled, this PR introduces a dedicated `ParallelStyle` to shard the
nn.LayerNorm/nn.Dropout/RMSNorm layers. We were mainly using a manual
distribute_module calls before when sharding the RMSNorm layer, but I
think we should have a dedicate TP API to easily shard those layers,
instead of user manually using DTensors.

I call this SequenceParallel, which might bring some confusion that we
technically "deprecated" a SequenceParallel style months ago. But this
time the SeuqenceParallel style is significantly different with the
previous ones (which used to shard two consecutive Linear layers). I
believe making it the right name is the first priority, instead of
worrying about the issue of reusing the old name

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

wanchaol committed:
[TP] Introduce Sequence Parallel Style for Laynorm/RMSNorm/Dropout

As titled, this PR introduces a dedicated `ParallelStyle` to shard the
nn.LayerNorm/nn.Dropout/RMSNorm layers. We were mainly using a manual
distribute_module calls before when sharding the RMSNorm layer, but I
think we should have a dedicate TP API to easily shard those layers,
instead of user manually using DTensors.

I call this SequenceParallel, which might bring some confusion that we
technically "deprecated" a SequenceParallel style months ago. But this
time the SeuqenceParallel style is significantly different with the
previous ones (which used to shard two consecutive Linear layers). I
believe making it the right name is the first priority, instead of
worrying about the issue of reusing the old name

[ghstack-poisoned]

wanchaol committed:
Update on "[TP] Introduce Sequence Parallel Style for Laynorm/RMSNorm/Dropout"


As titled, this PR introduces a dedicated `ParallelStyle` to shard the
nn.LayerNorm/nn.Dropout/RMSNorm layers. We were mainly using a manual
distribute_module calls before when sharding the RMSNorm layer, but I
think we should have a dedicate TP API to easily shard those layers,
instead of user manually using DTensors.

I call this SequenceParallel, which might bring some confusion that we
technically "deprecated" a SequenceParallel style months ago. But this
time the SeuqenceParallel style is significantly different with the
previous ones (which used to shard two consecutive Linear layers). I
believe making it the right name is the first priority, instead of
worrying about the issue of reusing the old name

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

wanchaol committed:
Update on "[TP] Introduce Sequence Parallel Style for Laynorm/RMSNorm/Dropout"


As titled, this PR introduces a dedicated `ParallelStyle` to shard the
nn.LayerNorm/nn.Dropout/RMSNorm layers. We were mainly using a manual
distribute_module calls before when sharding the RMSNorm layer, but I
think we should have a dedicate TP API to easily shard those layers,
instead of user manually using DTensors.

I call this SequenceParallel, which might bring some confusion that we
technically "deprecated" a SequenceParallel style months ago. But this
time the SeuqenceParallel style is significantly different with the
previous ones (which used to shard two consecutive Linear layers). I
believe making it the right name is the first priority, instead of
worrying about the issue of reusing the old name

[ghstack-poisoned]

awgu commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -242,6 +243,82 @@ def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
+
+    def _replicate_module_fn(self, name: str, module: nn.Module, device_mesh: DeviceMesh):
+        for p_name, param in module.named_parameters():
+            # simple replication with fixed ones_ init from LayerNorm/RMSNorm, which allow
```
Would it be safer to include a broadcast? I worry that the failure mode of this assumption would be divergent model replicas, which might be hard to debug.

wanchaol replied:
So for all our "claimed" supporting modules, this is not a issue:
* nn.LayerNorm, after init it always have weights are that ones, https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/normalization.py#L192
* similarly RMSNormPython also have ones weight
* nn.Dropout does not have parameters

So I think it's very safe to just mark the tensor as Replicate without doing communication during init, let me know if this make sense

wanchaol replied:
added a docstr to explicitly mention the `ones` init assumption

wanchaol committed:
Update on "[TP] Introduce Sequence Parallel Style for Laynorm/RMSNorm/Dropout"


As titled, this PR introduces a dedicated `ParallelStyle` to shard the
nn.LayerNorm/nn.Dropout/RMSNorm layers. We were mainly using a manual
distribute_module calls before when sharding the RMSNorm layer, but I
think we should have a dedicate TP API to easily shard those layers,
instead of user manually using DTensors.

I call this SequenceParallel, which might bring some confusion that we
technically "deprecated" a SequenceParallel style months ago. But this
time the SeuqenceParallel style is significantly different with the
previous ones (which used to shard two consecutive Linear layers). I
believe making it the right name is the first priority, instead of
worrying about the issue of reusing the old name

[ghstack-poisoned]

awgu commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -272,10 +272,14 @@ class SequenceParallel(ParallelStyle):
+
+    .. note:: SequenceParallel style assumes ones initialization if there are weights in the nn.Module (i.e.
+        ``nn.LayerNorm`` or ``RMSNorm``, and they by default have ones initialization). If you have custom
+        inits for the weights on those modules, need to broadcast the weights before/after parallelizing.
```
```suggestion
        inits for the weights on those modules, you need to broadcast the weights before/after parallelizing to ensure that they are replicated.
```

awgu commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -242,6 +243,86 @@ def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
 
+class SequenceParallel(ParallelStyle):
+    """
+    SequenceParallel replicate a compatible nn.Module parameters and run the sharded computation with the sharded
+    input that shards on the sequence dimension. Currently supporting ``nn.LayerNorm``, ``nn.Dropout`` and the
```
nit:
```suggestion
    SequenceParallel replicates a compatible ``nn.Module`` 's parameters and runs the sharded computation with input
    sharded on the sequence dimension the sharded. This currently supports``nn.LayerNorm``, ``nn.Dropout``, and the
```

awgu commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -242,6 +243,86 @@ def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
+    input that shards on the sequence dimension. Currently supporting ``nn.LayerNorm``, ``nn.Dropout`` and the
+    `RMSNorm python implementation <https://github.com/facebookresearch/llama/blob/main/llama/model.py#L34>`__
+
+    This style implements the operation that described in the paper
```
```suggestion
    This style implements the operation that is described in the paper
```

awgu commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -242,6 +243,86 @@ def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
+    This style implements the operation that described in the paper
+    `Reducing Activation Recomputation in Large Transformer Models <https://arxiv.org/abs/2205.05198>`__
+
+    Both the input and output of the nn.Module will be sharded on the sequence dimension.
```
```suggestion
    Both the input and output of the ``nn.Module`` will be sharded on the sequence dimension.
```

awgu commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -242,6 +243,86 @@ def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
+
+    Keyword Args:
+        sequence_dim (int, optional):
+            The sequence dimension of the input tensor for the nn.Module, this is used to annotate the input tensor to
```
```suggestion
            The sequence dimension of the input tensor for the ``nn.Module``, this is used to annotate the input tensor to
```

awgu commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -242,6 +243,86 @@ def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
+        elif isinstance(input_tensor, torch.Tensor):
+            return DTensor.from_local(input_tensor, device_mesh, [Shard(sequence_dim)], run_check=False)
+        else:
+            raise ValueError(f"expecting input of {mod} to be a torch.Tensor or DTensor")
```
nit:
```suggestion
            raise ValueError(f"expecting input of {mod} to be a torch.Tensor or DTensor but got {input_tensor")
```

awgu reviewed (approved):
LGTM! I just left some nits for the docstrings. We can check the docs render before landing.

tianyu-l reviewed (approved):
Great addition to the parallel styles which makes applying SP easy. Had some inline comments.

wanchaol committed:
Update on "[TP] Introduce Sequence Parallel Style for Laynorm/RMSNorm/Dropout"


As titled, this PR introduces a dedicated `ParallelStyle` to shard the
nn.LayerNorm/nn.Dropout/RMSNorm layers. We were mainly using a manual
distribute_module calls before when sharding the RMSNorm layer, but I
think we should have a dedicate TP API to easily shard those layers,
instead of user manually using DTensors.

I call this SequenceParallel, which might bring some confusion that we
technically "deprecated" a SequenceParallel style months ago. But this
time the SeuqenceParallel style is significantly different with the
previous ones (which used to shard two consecutive Linear layers). I
believe making it the right name is the first priority, instead of
worrying about the issue of reusing the old name

[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -242,6 +243,86 @@ def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
+    @staticmethod
+    def _prepare_input_fn(sequence_dim, mod, inputs, device_mesh):
+        input_tensor = inputs[0]
+        if isinstance(input_tensor, DTensor):
```
Yeah it's kind of hard to do the check for the reason you mentioned, also it's a bit challenging to make sure it works with torch.compile if we add this metadata check. I think we can leave it as a follow up later

wanchaol commented:
@pytorchbot merge
